README.md

```diff
@@ -0,0 +1,173 @@
+  ```
+  # Per convention -- indicates the end of command options
+  $ cli start -- bash -c "echo Hello World"
+  Started job with ID: a2f3d7865b4e
+  ```
+
+* **Stopping a Job:**  
+  ```
+  $ cli stop 4d5c9fa91db4
+  ```
+
+* **Query Job Status:**
+  ```
+  $ cli status 4d5c9fa91db4
+  running, completed, failed or terminated 
+  ```
+
+* **Viewing Job Output:**  
+  ```
+  $ cli log 4d5c9fa91db4
+  ```
+
+* **Streaming Job Ouput:**
+  ```
+  $ cli log -f 4d5c9fa91db4
+  ```
+
+#### **Future CLI Improvements**
+
+It's worth mentioning that our CLI UX is basic and we would probably want to build something better in production. To improve our user CLI experience we could add:
+
+* **Job Listing**
+  ```
+  $ cli ls
+  ```
+
+* **Job Naming**
+  ```
+  $ cli start -n hello_world -- echo Hello World 
+  ```
+
+For now, we'll leave these two improvements for the future to reduce our implementation scope. 
+
+### **2. Library**
+
+#### **Interface**
+
+The library will expose the following functions as the primary interface:
+
+```go
+type struct Job {
+  id
+  <internal_fields>
+}
+type struct LogReader {
+  <internal_fields>
+}
+
+// Managing jobs and queueing job status
+StartJob(command string, args ...string) (*Job, error)
+func (job *Job) Stop() error
+func (job *Job) GetStatus() string
+
+// Reading job output
+func (job *Job) GetNewLogReader() (*LogReader, error)
+func (reader *LogReader) GetNextLineBlocking()
+func (reader *LogReader) GetNextLineNonBlocking()
+```
+#### **Job Life Cycle**
+
+At a high level when creating a job we:
+1. Configure cgroups and set hardcode resource limits 
+2. Setup isloation for PID, mount, and networking namespaces.
+3. Initialize job command in a new process
+4. Run go routines to monitor job completion and failure and capture stdout/stderr output into a job log
+5. Run job process until completion/failure
+6. Run namespace and cgroup cleanup 
+
+#### **Multi-Client Streaming**
+
+There can multiple clients subscribed to subscribed a job's output. Each individual client simply has to create their own job reader through `GetNewLogReader` and call the `GetNextLine` functions to consume log lines at their own pace. 
+
+### **3. gRPC Server**
+
+#### **API Interface**
+
+Here are the proposed proto definitions for the API:
+
+```proto
+service JobService {
+    rpc Start(JobStartRequest) returns (JobStartResponse);
+    rpc Stop(JobStopRequest) returns (JobStopResponse);
+    rpc QueryStatus(JobQueryRequest) returns (JobInfo);
+    rpc SubscribeOutput(JobSubscriptionRequest) returns (stream JobOutputResponse); // Updated for streaming
+}
+
+message JobStartRequest {
+    string command = 1;
+    repeated string args = 2;
+}
+
+message JobStartResponse {
+    string jobId = 1;
+}
+
+message JobStopRequest {
+    string jobId = 1;
+}
+
+message JobStopResponse {
+    string status = 1; // e.g., "stopped", "not found", "error"
+    string message = 2; // e.g., "Job successfully stopped", "Job not found", etc.
```

Comment:
these two fields seem like they're basically the same, no?

Reply:
Yes fair enough. I'll remove this field.